src/core/coap/coap_secure.cpp

```diff
@@ -351,6 +366,35 @@ otError CoapSecure::HandleDtlsSend(const uint8_t *aBuf, uint16_t aLength, uint8_
     return error;
 }
 
+void CoapSecure::HandleTransmit(Tasklet &aTasklet)
+{
+    aTasklet.GetOwner<CoapSecure>().HandleTransmit();
+}
+
+void CoapSecure::HandleTransmit(void)
+{
+    otError  error   = OT_ERROR_NONE;
+    Message *message = mTransmitQueue.GetHead();
+
+    VerifyOrExit(message != NULL);
+    mTransmitQueue.Dequeue(*message);
+
+    if (mTransmitQueue.GetHead() != NULL)
+    {
+        mTransmitTask.Post();
+    }
+
+    SuccessOrExit(error = GetDtls().Send(*message, message->GetLength()));
```

Comment:
One potential downside is any error returned by `Dtls::Send()` will not propagate back to the client/higher-layer. Thoughts?

Reply:
Yes, this is the downside. However, I think we benefit memory cost from this PR, both stack depth and message buffers. How about we enhance the error handling here? For example, when no memory, we may defer again, when connection is aborted, we may abort the current message. Since this PR only affects CoAP, which has retransmission, so maybe it's not critical to propagate all errors up to client or server. Thoughts?

Reply:
Thinking about this again, I'm fine with how you have it now. If it becomes a problem, we can consider addressing it at that time. But maybe worth having a log message at notice level when an error does occur.

Reply:
Added, thanks for the suggest.